draft-wood-key-consistency.md

```diff
@@ -148,55 +158,132 @@ to ensure consistency here might include, though are not limited to:
-
-If the database is small and users query it infrequently, retrieval techniques based on PIR may be viable.
-
-## Minimum Validity Periods
+## Database Discovery {#external-db-based}
+
+In this model, servers publish keys in an external database and clients fetch keys from the database, as shown below.
+
+~~~
+┌──────────┐
+│          │
+│  Client  ├───────────┐
+│          │           │
+└──────────┘           │
+                       │
+┌──────────┐         ┌─▼────────┐      ┌─────────┐
+│          │         │          │      │         │
+│  Client  ├─────────► Database ◄──────┤ Server  │
+│          │         │          │      │         │
+└──────────┘         └─▲────────┘      └─────────┘
+     x                 │
+     x                 │
+     x                 │
+┌──────────┐           │
+│          │           │
+│  Client  ├───────────┘
+│          │
+└──────────┘
+~~~
+{: #fig-disc-database title="Database Discovery Example"}
+
+The database is expected to have a table that asserts mappings between server names and keys. Examples
+of such databases are as follows:
+
+- An append-only, audited table similar to that of Certificate Transparency {{!RFC6962}}. The log is operated and
+  audited in such a way that the contents of the log are consistent for all users. Any reliant system
+  which depends on this type of KCCS requires the log be audited or users have some other mechanism for
+  checking their view of the log state (gossiping). However, this type of system does not ensure proactive
+  security against malicious servers unless log participants actively check log proofs. This requirement
+  may impede deployment in practice, given that no web browser checks SignedCertificateTimestamps before
+  using (accepting as valid) a corresponding TLS certificate.
+
+- A consensus-based table whose assertions are created by a coalition of entities that periodically agree on
+  the correct binding of server names and key material. In this model the agreement is achieve via a consensus
+  protocol, but the specific consensus protocol is dependent on the implementation. For privacy, users should
+  either download the entire database and query it locally, or remotely query the database using a private
+  information retrieval (PIR) protocol. In the case where the database is downloaded locally, it should be
+  considered stale and re-fetch periodically, as well.
+
+  When the entire database is downloaded, this model is appropriate in small scale deployments where the
+  number of bindings in the database is much smaller than the number of users of the reliant system. In
+  a reliant system with a large user base, this model imposes bandwidth costs on each user that may be
+  impractical. In larger scale deployments, the short-comings of this model may be similar to log-based
+  tables.
+
+- A Private Information Retrieval (PIR) table. If the database is small and users query it infrequently,
```

Comment:
I'm not sure how this is different from the previous example?

Reply:
Fair point -- PIR is more about how one *accesses* the table, rather than the contents of the table itself. We could maybe pull PIR out into its own discovery mechanism, then? Not sure.

Reply:
The original goal was describing how to safely query a database either locally or remotely, where the client does not leak/reveal to a third-party which entry they want from the database. These two paragraphs were focused on the trade-off of the two solutions. However, in both of these designs, an entry includes some guarantee that it was considered "correct" by whatever-consensus-protocol was used (maybe multiple signatures, or a threshold signature, or another mechanism entirely).

With that being said, "a third-party database" could be broken out into a separate discovery mechanism, like "Indirect Discovery" :) And I think this would have roughly the same criteria as "Direct Discovery".

Reply:
How is "indirect discovery" different from "database discovery"? (I like your term *a lot* more!)

Reply:
Ah! That's a really good question. I guess I forgot these were all combined under a single "Database Discovery" model, and I think they are not different :slightly_smiling_face:

I guess I could be clearer by saying PIR fits under "consensus-based" and "trusted third-party"? For consensus-based, we have the current text. For "third-party" we have basically "single-proxy discovery" as above except the "proxy" only provides the key and then the client connects to the destination server through another path (directly or another proxy or however else).

Reply:
Yeah, that might work. Let me give that a shot.

Reply:
@sysrqb okay, rewritten somewhat. The PIR text now applies to a generic database, as intended (I think). What do you think?